components/zip321-parse/src/lib.rs

```diff
@@ -0,0 +1,924 @@
+) -> Result<IndexedParam, String> {
+    let payment_index = match iopt {
+        Some(istr) => istr.parse::<usize>().map(Some).map_err(|e| e.to_string()),
+        None => Ok(None),
+    }?;
+
+    let param = match name {
+        "address" => Ok(Param::Addr(value.to_owned())),
+        "amount" => parse_amount(value)
+            .map_err(|e| e.to_string())
+            .map(|(_, (coins, zatoshis))| Param::Amount { coins, zatoshis }),
+
+        "label" => percent_decode(value.as_bytes())
+            .decode_utf8()
+            .map(|s| Param::Label(s.into_owned()))
+            .map_err(|e| e.to_string()),
+
+        "message" => percent_decode(value.as_bytes())
+            .decode_utf8()
+            .map(|s| Param::Message(s.into_owned()))
+            .map_err(|e| e.to_string()),
+
+        "memo" => Memo::try_from_base64(value)
+            .map(|m| Param::Memo(m))
+            .map_err(|e| format!("Decoded memo was invalid: {e:?}")),
+        other if other.starts_with("req-") => {
+            Err(format!("Required parameter {other} not recognized"))
+        }
+
+        other => percent_decode(value.as_bytes())
+            .decode_utf8()
+            .map(|s| Param::Other(other.to_string(), s.into_owned()))
+            .map_err(|e| e.to_string()),
+    }?;
+
+    Ok(IndexedParam {
+        param,
+        payment_index: payment_index.unwrap_or(0),
+    })
+}
+
+/// Converts an vector of [`Param`] values to a [`Payment`].
+///
+/// This function performs checks to ensure that the resulting [`Payment`] is structurally
+/// valid; for example, a request for memo contents may not be associated with a
+/// transparent payment address.
+pub fn to_payment(vs: Vec<Param>, i: usize) -> Result<Payment> {
+    let recipient_address = vs
+        .iter()
+        .find_map(|v| match v {
+            Param::Addr(a) => Some(a.clone()),
+            _otherwise => None,
+        })
+        .context(RecipientMissingSnafu { idx: i })?;
+
+    let mut payment = Payment {
+        recipient_address,
+        amount_coins: 0,
+        amount_zatoshis: 0,
+        memo: None,
+        label: None,
+        message: None,
+        other_params: vec![],
+    };
+
+    for v in vs {
+        match v {
+            Param::Amount { coins, zatoshis } => {
+                payment.amount_coins = coins;
+                payment.amount_zatoshis = zatoshis;
+            }
+            Param::Memo(m) => {
+                payment.memo = Some(m);
+            }
+            Param::Label(m) => payment.label = Some(m),
+            Param::Message(m) => payment.message = Some(m),
+            Param::Other(n, m) => payment.other_params.push((n, m)),
+            _otherwise => {}
+        }
+    }
+
+    Ok(payment)
+}
+
+/// A single payment being requested.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Payment {
+    // TODO(schell): changed
+    /// The address to which the payment should be sent.
+    recipient_address: String,
```

Comment:
The type here changed from `ZcashAddress` to `String`. 

The `zip321::Payment` type wraps this one and adds `ZcashAddress` on top, validating it.